app/controllers/uploads_controller.rb

```diff
@@ -270,6 +301,186 @@ def complete_external_upload
+      return render_json_error(I18n.t("upload.failed"), status: 422)
+    end
+
+    upload_stub = ExternalUploadStub.create!(
+      key: multipart_upload[:key],
+      created_by: current_user,
+      original_filename: file_name,
+      upload_type: upload_type,
+      external_upload_identifier: multipart_upload[:upload_id],
+      multipart: true,
+      filesize: file_size
+    )
+
+    render json: {
+      external_upload_identifier: upload_stub.external_upload_identifier,
+      key: upload_stub.key,
+      unique_identifier: upload_stub.unique_identifier
+    }
+  end
+
+  def batch_presign_multipart_parts
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+
+    part_numbers = params.require(:part_numbers)
+    unique_identifier = params.require(:unique_identifier)
+
+    RateLimiter.new(
+      current_user, "batch-presign-#{unique_identifier}", BATCH_PRESIGN_RATE_LIMIT_PER_MINUTE, 1.minute
+    ).performed!
+
+    part_numbers = part_numbers.map(&:to_i).map do |part_number|
+      validate_part_number(part_number)
+      part_number
+    end
+
+    external_upload_stub = ExternalUploadStub.find_by(
+      unique_identifier: unique_identifier, created_by: current_user
+    )
+    return render_404 if external_upload_stub.blank?
+
+    if !multipart_upload_exists?(external_upload_stub)
+      return render_404
+    end
+
+    presigned_urls = {}
+    part_numbers.each do |part_number|
+      presigned_urls[part_number] = Discourse.store.presign_multipart_part(
+        upload_id: external_upload_stub.external_upload_identifier,
+        key: external_upload_stub.key,
+        part_number: part_number
+      )
+    end
+
+    render json: { presigned_urls: presigned_urls }
+  end
+
+  def validate_part_number(part_number)
+    if part_number.zero? || part_number.negative? || !part_number.between?(1, 10000)
+      raise Discourse::InvalidParameters.new(
+        "Part numbers should be a list of numbers between 1 and 10000"
+      )
+    end
+  end
+
+  def multipart_upload_exists?(external_upload_stub)
+    begin
+      Discourse.store.list_multipart_parts(
+        upload_id: external_upload_stub.external_upload_identifier, key: external_upload_stub.key
+      )
+    rescue Aws::S3::Errors::NoSuchUpload => err
+      debug_upload_error(err, "upload.external_upload_not_found", { additional_detail: "path: #{external_upload_stub.key}" })
+      return false
+    end
+    true
+  end
+
+  def abort_multipart
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+
+    external_upload_identifier = params.require(:external_upload_identifier)
+
+    RateLimiter.new(
+      current_user, "abort-multipart-upload", ABORT_MULTIPART_RATE_LIMIT_PER_MINUTE, 1.minute
+    ).performed!
+
+    external_upload_stub = ExternalUploadStub.find_by(
+      external_upload_identifier: external_upload_identifier,
+      created_by: current_user
+    )
+    return render_404 if external_upload_stub.blank?
```

Comment:
I think we should just make this action idempotent. If the `external_upload_stub` is not found, it is safe to mean that the record has been destroyed so we can just return a `success_json`.